add blog post about kms plugins (#72)
sigstore/sigstore-blog: content/kms-plugins.md

+++
title = "KMS Plugins for Sigstore"
author = "Ramon Petgrave (@ramonpetgrave64)"
date = "2025-02-19"
draft = false
type = "post"
tags = ["sigstore","kms", "plugin"]
+++

> Cosign and private deployments of Fulcio, and Rekor can use an external KMS provider. We currently have built-in support for AWS, Azure, GCP, and Hashicorp Vault KMSs. This has been a challenge for customers that require alternative or custom KMS solutions.

Reviewer: ```suggestion
Cosign and private deployments of Fulcio and Rekor can use a KMS-managed key for signing artifacts. We currently have built-in support for AWS, Azure, Google Cloud Platform, and Hashicorp Vault KMSs. This has been a challenge for customers that require alternative or custom KMS solutions.
```

To enable such use-cases, we have implemented a new [plugin system](https://github.com/sigstore/sigstore/tree/main/pkg/signature/kms/cliplugin) for alternate KMS providers. Organizations can independently and privately develop & distribute their plugins without needing downstream updates to libraries to support additional KMS providers as build-time dependencies.

### How it Works

> The Sigstore KMS plugins will be separate, installable programs. Users will invoke Cosign specifying their plugin with the key ref like `cosign sign … —key mykms://my-key-id`, where `mykms` refers to a plugin program `sigstore-kms-mykms` located on the system PATH. The existing key refs for the existing built-in KMS providers will continue to work the same and take precedence.

Reviewer: ```suggestion
The Sigstore KMS plugins will be separate, installable programs. Users will invoke Cosign specifying their plugin with the key ref like `cosign sign … -- key mykms://my-key-id`, where `mykms` refers to a plugin program `sigstore-kms-mykms` located on the system PATH. The existing key references for the existing built-in KMS providers will continue to work the same and take precedence.
```

+++
title = "KMS Plugins for Sigstore"
author = "Ramon Petgrave (@ramonpetgrave64)"

> date = "2025-02-19"

Reviewer 2: can we update this to indicate 4-4-2025 and we'll publish this tomorrow?

Author: Changed to Monday Apr 7, in case we want to mention another provider.
 - https://github.com/sigstore/sigstore-blog/pull/72/files#r2027570830

draft = false
type = "post"
tags = ["sigstore","kms", "plugin"]
+++

Cosign and private deployments of Fulcio and Rekor can use a KMS-managed key for signing artifacts. We currently have built-in support for AWS, Azure, Google Cloud Platform, and Hashicorp Vault KMSs. This has been a challenge for customers that require alternative or custom KMS solutions.

> To enable such use-cases, we have implemented a new [plugin system](https://github.com/sigstore/sigstore/tree/main/pkg/signature/kms/cliplugin) for alternate KMS providers. Organizations can independently and privately develop & distribute their plugins without needing downstream updates to libraries to support additional KMS providers as build-time dependencies.

Reviewer 3: another advantage is that people don't need to build and maintain their own forks just to support a new KMS provider

Author: Added

> [Alibaba KMS](https://github.com/mozillazg/sigstore-kms-alibabakms) support is now possible thanks to new plugin by community member @mozillazg.

Reviewer 2: ```suggestion
[Alibaba KMS](https://github.com/mozillazg/sigstore-kms-alibabakms) support is now possible thanks to a new plugin by community member @mozillazg.
```

> 

Author: @zosocanuck would you like a mention for the [venafi plugin](https://github.com/Venafi/sigstore-kms-venafi)?

### How it Works

> The Sigstore KMS plugins will be separate, installable programs. Users will invoke Cosign specifying their plugin with the key ref like `cosign sign … -- key mykms://my-key-id`, where `mykms` refers to a plugin program `sigstore-kms-mykms` located on the system PATH. The existing key references for the existing built-in KMS providers will continue to work the same and take precedence.

Reviewer 2: ```suggestion
The out-of-tree Sigstore KMS plugins will be separate, installable programs. Users will invoke Cosign specifying their plugin with the key ref like `cosign sign … -- key mykms://my-key-id`, where `mykms` refers to a plugin program `sigstore-kms-mykms` located on the system PATH. The existing key references for the existing built-in KMS providers will continue to work the same and take precedence.
```

> The out-of-tree Sigstore KMS plugins will be separate, installable programs. Users will invoke Cosign specifying their plugin with the key ref like `cosign sign … -- key mykms://my-key-id`, where `mykms` refers to a plugin program `sigstore-kms-mykms` located on the system PATH. The existing key references for the existing built-in KMS providers will continue to work the same and take precedence.

Reviewer 3: ```suggestion
The out-of-tree Sigstore KMS plugins will be separate, installable programs. Users will invoke Cosign specifying their plugin with the key ref like `cosign sign … --key mykms://my-key-id`, where `mykms` refers to a plugin program `sigstore-kms-mykms` located on the system PATH. The existing key references for the existing built-in KMS providers will continue to work the same and take precedence.
```

I think this space doesn't belong?

### Plugin Development

Like [kubectl](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/#writing-kubectl-plugins) plugins, Sigstore KMS plugins can be written in any language on any OS that can write programs that accept CLI arguments and stdin, and output to stdout. Plugin authors see more [details](https://github.com/sigstore/sigstore/tree/main/pkg/signature/kms/cliplugin) and can follow our [example](https://github.com/sigstore/sigstore/blob/main/test/cliplugin/localkms) plugin program as a guide.

> Developers that use sigstore/sigstore as a library can also use plugins with the `kms.Get()` function.

Reviewer 2: ```suggestion
Developers that use [sigstore/sigstore](https://github.com/sigstore/sigstore) as a library can also use plugins with the `kms.Get()` function.
```
